man/segyio-cath.1
.TH SEGYIO-CATH 1
.SH NAME
segyio-cath \- Concatenate SEG-Y text headers and print on the standard output
.SH SYNPOSIS
.B segyio-cath
[\fIOPTION\fR]...
[\fIFILE\fR]...
.SH DESCRIPTION
.B segyio-cath
Concatenate the SEG-Y text headers in FILEs.

.PP
By default, only the non-extended header is printed, which is equivalent to
--num 0.

.PP
Mandatory arguments to long options are mandatory for short options too.

.SH OPTIONS
.TP
.BR \-n ", " \-\-num =\fINUM\fR
the textual header to show, starts at 0

.TP
.BR \-a ", " \-\-all
print all headers

.TP
.BR \-s ", " \-\-strict
abort if a header or file is not found
primarily meant for shell scripts

.TP
.BR \-S ", " \-\-nonstrict
ignore missing headers
this is the default behaviour

.TP
.BR \-\-version
output version information and exit

.TP
.BR \-\-help
display this help and exit

.SH COPYRIGHT
Copyright © Statoil ASA. License LGPLv3+: GNU LGPL version 3 or later <http://gnu.org/licenses/lgpl.html>.

.PP
This is free software: you are free to change and redistribute it.  There is NO WARRANTY, to the extent permitted by law.

.SH SEE ALSO
segyio-catb(1), segyio-catr(1)


changelog.md
# 1.3.1
* fix a typo in segyio-crop --version
* some building improvements

# 1.3.0
* segyio is now meant to be used as proper versions, not trunk checkouts.
  changelogs from now on will be written when new versions are released, not on
  a monthly basis.
* Minor typo fixes in segyio-cath help
* Applications now come with man pages
* `header` modes handle negative indexing
* `header.update` handle any key-value iterable
* New application: segyio-catr for printing trace headers
* New application: segyio-catb for printing the binary header
* New application: segyio-crop for copying sub cubes

# 2017.06
* seismic unix-style aliases are available for python in the segyio.su
  namespace
* segyio has learned how to calculate the rotation of the cube
  (segy_rotation_cw and segyio.tools.rotation)
* The python header objects behave more as dicts as expected
* The new program segyio-cath is added, a cat-like program that concatenate
  text headers.
* Infrastructure improvements
* segyio for python is available via pypi (pip)
* segyio is now meant to be consumed with binary downloads of versions, but
  with a rapid release cycle. Releases within a major release will be backwards
  compatible
* Shared linking of the python extension is considered deprecated on Windows.

# 2017.05
* Requirements for the shape of the right hand side of `f.trace[:] =`
  expressions are relaxed and accepts more inputs
* C interface slightly cleaned up for C99 compliance
* C library can reason on arbitrary header words for offsets, not just 37

# 2017.04
* Examples in the readme
* Delay recording time (t0) is interpreted as milliseconds, not microseconds
* Some minor optimisations
* segy_mmap is more exception safe
* The applications warn if mmap fails
* Support for static analysis with cppcheck
* The statoil/pycmake repo is used for python integration in cmake
* Minor milli/microsecond bugfixes in mex bindings
* tools.wrap added for printing textual headers to screen or file

# 2017.03
* Float conversions (ibm <-> native) has been optimised and is much faster
* `segy_binheader_size` returns signed int like its friends
* `sample_interval` steps are now floats
* Multiple internal bug fixes
* Some buffer leaks are plugged
* segyio has learned to deal with files without good geometry.
  If `strict = False` is passed to segyio.open, and a file is without
  well-sorted inlines/crosslines, open will return a file handle, but with
  geometry-dependent modes disabled
* `file.samples` returns a list of samples, not number of samples
* Readme has been improved.
* `trace[int]` is more robust w.r.t. inputs
* A new mode has been added; gather. gather depends on a good geometry, and its
  getitem `[il, xl, slice(offsets)]` returns all offsets for an
  inline/crossline intersection

# 2017.02
* segyio has learned to deal with large files (>4G) on more platforms
* segyio can read quickly attributes (trace header words) over the full file
* python can tell the fast and slow directions apart
* Reading depth slices is much faster
* tools.collect for gathering read samples into a single numpy ndarray
* tools.cube for easily reading a full cube
* tools.native for fast third-party segy-to-native float conversion
* File opening in binary mode is now enforced
* Data types have been overhauled (prefer signed integers)
* Enumerations have been SEGY prefixed to reduce collisions
* Building shared libs can be switched on/off on cmake invocation
* Makefiles and CI overhauls

# 2017.01
* Matlab has learned about prestack files
* Reading traces in matlab no longer fails when not reading the whole file
* Matlab argument keys have been renamed

# 2016.11
* Fixed some condtions where a failed write would corrupt trace data
* Fixed a memory leak bug
* VERSION string added to python
* Experimental memory-mapped file support
* Line-oriented C functions are offset aware
* Python offset property exposes offset numbers, not just count
* Support for pre-stack files, with new subindexing syntax, line[n, offset]
* Improved python repl (shell, read-eval-print-loop) support
* The widgets have color- and layout selectors

# 2016.10
* Matlab tests can optionally be turned off
* The application Segyviewer is embeddable and provided by the segyview sub
  library
* libcwrap has been replaced fully by the python C api
* OS X and experimental Windows support
* A new sub mode for traces, raw, for eager reading of trace data


applications/segyio-cath.c
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <getopt.h>
#include <unistd.h>

#include "apputils.c"
#include <segyio/segy.h>

static int help() {
    puts( "Usage: segyio-cath [OPTION]... [FILE]...\n"
          "Concatenate the textual header(s) from FILE(s) to standard output.\n"
          "\n"
          "-n, --num        the textual header to show, starts at 0\n"
          "-a, --all        all textual headers\n"
          "-s, --strict     abort if a header or file is not found\n"
          "                 primarily meant for shell scripts\n"
          "-S, --nonstrict  ignore missing headers\n"
          "                 this is the default behaviour\n"
          "    --version    output version information and exit\n"
          "    --help       display this help and exit\n"
          "\n"
          "By default, only the non-extended header is printed, which is\n"
          "equivalent to --num 0\n"
        );
    return 0;
}

static int printversion() {
    printf( "segyio-cath (segyio version %d.%d)\n", segyio_MAJOR, segyio_MINOR );
    return 0;
}

static int ext_headers( segy_file* fp ) {
    char binary[ SEGY_BINARY_HEADER_SIZE ];
    int err = segy_binheader( fp, binary );

    if( err ) return -1;

    int32_t ext;
    err = segy_get_bfield( binary, SEGY_BIN_EXT_HEADERS, &ext );
    if( err ) return -2;
    return ext;
}

static void print_header( const char* header ) {
    for( int line = 0, ch = 0; line < 40; ++line ) {
        for( int c = 0; c < 80; ++c, ++ch ) putchar( header[ ch ] );
        putchar( '\n' );
    }
}

int main( int argc, char** argv ) {

    static int all = false;
    static int strict = false;
    static int version = 0;
    static struct option long_options[] = {
        { "num",        required_argument,    0,        'n' },
        { "all",        no_argument,          &all,      1  },
        { "strict",     no_argument,          &strict,   1  },
        { "nonstrict",  no_argument,          &strict,   0  },
        { "version",    no_argument,          &version,  1  },
        { "help",       no_argument,          0,        'h' },
        { 0, 0, 0, 0 }
    };

    int num_alloc_sz = 32;
    int* num   = calloc( sizeof( int ), num_alloc_sz );
    int num_sz = 0;

    while( true ) {
        int option_index = 0;
        int c = getopt_long( argc, argv, "n:asS",
                             long_options, &option_index );

        if( c == -1 ) break;

        char* endptr;
        switch( c ) {
            case  0:  break;
            case 'h': return help();
            case 's': strict  = 1; break;
            case 'S': strict  = 0; break;
            case 'a': all     = 1; break;
            case 'n':
                if( version ) break;

                if( num_sz == num_alloc_sz - 1 ) {
                    num_alloc_sz *= 2;
                    int* re = realloc( num, num_alloc_sz * sizeof( int ) );
                    if( !re ) return errmsg( errno, "Unable to alloc" );
                    num = re;
                }

                num[ num_sz ] = strtol( optarg, &endptr, 10 );
                if( *endptr != '\0' )
                    return errmsg( EINVAL, "num must be an integer" );
                if( num[ num_sz ] < 0 )
                    return errmsg( EINVAL, "num must be non-negative" );
                break;

            default: return help();
        }
    }

    if( version ) return printversion();

    char header[ SEGY_TEXT_HEADER_SIZE + 1 ] = { 0 };

    for( int i = optind; i < argc; ++i ) {
        segy_file* fp = segy_open( argv[ i ], "r" );

        if( !fp ) fprintf( stderr, "segyio-cath: %s: No such file or directory\n",
                           argv[ i ] );

        if( !fp && strict ) return errmsg( errno, NULL );
        if( !fp ) continue;

        if( num_sz == 0 ) num_sz = 1;

        const int exts = ext_headers( fp );
        if( exts < 0 )
            return errmsg( errno, "Unable to read binary header" );

        if( all ) {
            /* just create the list 0,1,2... as if it was passed explicitly */
            if( exts >= num_alloc_sz ) {
                num_alloc_sz = exts * 2;
                int* re = realloc( num, num_alloc_sz * sizeof( int ) );
                if( !re ) return errmsg( errno, "Unable to alloc" );
                num = re;
            }

            num_sz = exts + 1;
            num[ 0 ] = 0;
            for( int j = 0; j < exts; ++j )
                num[ j + 1 ] = j;
        }

        for( int j = 0; j < num_sz; ++j ) {
            if( strict && ( num[ j ] < 0 || num[ j ] > exts ) )
                return errmsg( EINVAL, "Header index out of range" );
            if( num[ j ] < 0 || num[ j ] > exts ) continue;

            int err = num[ j ] == 0
                ? segy_read_textheader( fp, header )
                : segy_read_ext_textheader( fp, num[ j ] - 1, header );

            if( err != 0 ) return errmsg( errno, "Unable to read header" );
            print_header( header );
        }

        segy_close( fp );
    }

    free( num );
    return 0;
}
